mrmustard/lab_dev/transformations/phasenoise.py

```diff
@@ -0,0 +1,91 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The class representing a Phase noise channel.
+"""
+
+from __future__ import annotations
+from typing import Sequence
+from mrmustard import math
+from mrmustard.physics.representations import Representation
+from mrmustard.physics.ansatz.array_ansatz import ArrayAnsatz
+from mrmustard.lab_dev.circuit_components import CircuitComponent
+from .base import Channel
+from ..utils import make_parameter
+import numpy as np
+
+__all__ = ["PhaseNoise"]
+
+
+class PhaseNoise(Channel):
+    r"""
+    The Phase noise channel.
+
+    This class represents the application of a random phase. The distributiuon of the phase
+    is assumed to be a Gaussian with mean zero, and standard deviation `sigma`.
+
+    Args:
+        modes: The modes the channel is applied to
+        sigma: the standard deviation of the random Gaussian noise (could be provided as a list, if the channel acts on more than one mode)
+
+    ..details::
+        The Fock representation is connected to the Fourier coefficients of the distribution.
+
+    """
+
+    short_name = "P~"
+
+    # randomized : bool
+    def __init__(
+        self,
+        modes: Sequence[int],
+        phase_stdev: float | Sequence[float],
+        phase_stdev_trainable: bool = False,
+        phase_stdev_bounds: tuple[float | None, float | None] = (0.0, None),
+    ):
+        super().__init__(name="PhN")
+        self._add_parameter(
+            make_parameter(phase_stdev_trainable, phase_stdev, "phase_stdev", phase_stdev_bounds)
+        )
+        self._representation = self.from_ansatz(
+            modes_in=modes, modes_out=modes, ansatz=None
+        ).representation
+
+    def __custom_rrshift__(self, other):
+        r"""
+        Custom rrshift
+        """
+        dm_op = getattr(other, "dm", None)
+        if callable(dm_op):
+            dm_state = other.dm()
+            array = math.asnumpy(dm_state.fock_array())
+            for mode in self.modes:
+                for count, _ in enumerate(np.nditer(array)):
+                    idx = np.zeros(len(array.shape))
+                    temp = count
+                    for l in range(len(idx)):
+                        idx[-1 - l] = temp % array.shape[-1 - l]
+                        temp = temp // array.shape[-1 - l]
+                    array_index = tuple(idx.astype(int))
+                    array[array_index] *= np.exp(
+                        -0.5
+                        * (idx[mode] - idx[other.n_modes + mode]) ** 2
+                        * self.phase_stdev.value**2
+                    )
+            return CircuitComponent(
+                Representation(ArrayAnsatz(array, False), dm_state.wires), self.name
+            )
+        else:
+            raise ValueError("The PhaseNoise object can only be applied to a DM or a Ket.")
```

Comment:
```suggestion
            raise ValueError("PhaseNoise can only be applied to a DM or a Ket.")
```

Reply:
Maybe there's an issue with this. If what started as a Ket or DM contracts with other stuff it will no longer be a Ket or a DM, but it's still correct to apply the PhaseNoise to those wires.

Reply:
Doesn't the rshift of Ket and DM guarantee that anytime a state is contracted with a transformation it returns a state?

Reply:
I guess it is fixed now?

Reply:
What I'm saying is that if the object that initially is a Ket or DM contracts with something else before it contracts with PhaseNoise, it can be that the resulting object isn't a Ket or a DM and the contraction with PhaseNoise would fail, despite being the correct thing to do next.

Reply:
Well, now it works on those cases Filippo, no?